libraries/AP_ExternalAHRS/AP_ExternalAHRS_AdvancedNavigation.cpp

```diff
@@ -0,0 +1,792 @@
+            },
+            AN_PERIOD{
+                id: AN_PACKET_ID_VELOCITY_STANDARD_DEVIATION,
+                packet_period: (uint32_t) 1.0e3 / _current_rate
+            },
+            AN_PERIOD{
+                id: AN_PACKET_ID_RAW_SENSORS,
+                packet_period: (uint32_t) 1.0e3 / _current_rate
+            },
+            AN_PERIOD{
+                id: AN_PACKET_ID_RAW_GNSS,
+                packet_period: (uint32_t) 1.0e3 / _current_rate
+            },
+            AN_PERIOD{
+                id: AN_PACKET_ID_SATELLITES,
+                packet_period: (uint32_t) 1.0e3 / _current_rate
+            }
+        }
+    };
+
+    AN_PACKET packet;
+    // load the AN_PACKETS_PERIOD Into the payload.
+    packet.payload.packets_period = periods;
+    packet.update_checks(AN_PACKET_ID_PACKET_PERIODS, (5*sizeof(AN_PERIOD)) + 2);
+
+    // Check for space in the tx buffer
+    if (_uart->txspace() < packet.packet_size()) {
+        return false;
+    }
+    _uart->write(packet.raw_pointer(), packet.packet_size());
+
+    return true;
+}
+
+/*
+ * Function that returns the gps capability of the connected AdNav device.
+ */
+bool AP_ExternalAHRS_AdvancedNavigation::get_gnss_capability(void) const
+{
+    switch (_device_id) {
+    case device_id_orientus:
+    case device_id_orientus_v3:
+    case device_id_air_data_unit:
+    case device_id_motus:
+    case device_id_certus_mini_a:
+        return false;
+        break;
+    default:
+        return true;
+        break;
+    }
+}
+
+/*
+ * Function that returns the barometric capability of the connected AdNav device.
+ */
+bool AP_ExternalAHRS_AdvancedNavigation::get_baro_capability(void) const
+{
+    switch (_device_id) {
+    case device_id_air_data_unit:
+    case device_id_orientus:
+    case device_id_orientus_v3:
+    case device_id_gnss_compass:
+    case device_id_certus_mini_a:
+        return false;
+        break;
+    case device_id_motus:
+        // Motus versions prior to 2.3 didn't have a barometer enabled.
+        if (_hardware_rev < 2300) {
+            return false;
+        }
+        break;
+    default:
+        break;
+    }
+    return true;
+}
+
+void AP_ExternalAHRS_AdvancedNavigation::handle_packet()
+{
+    // get current time
+    uint32_t now = AP_HAL::millis();
+    _last_pkt_ms = now;
+
+    // Update depending on received packet.
+    switch (_msg.packet.id) {
+    case AN_PACKET_ID_DEVICE_INFO: {
+        _last_device_info_pkt_ms = now;
+        _device_id = _msg.packet.payload.device_info.device_id;
+        _hardware_rev = _msg.packet.payload.device_info.hardware_revision;
+        break;
+    }
+    case AN_PACKET_ID_SYSTEM_STATE: {
+        _last_state_pkt_ms = now;
+        _system_status = _msg.packet.payload.system_state.system_status.r;
+        _filter_status = _msg.packet.payload.system_state.filter_status.r;
+        {
+            WITH_SEMAPHORE(state.sem);
+
+            state.accel = Vector3f{
```

Comment:
I wonder if we should just keep around the most recent instance of each of these packets, rather than copying stuff out fieldwise.  How much information would we be storing that we don't need to?  To a *certain* extent flash is more precious to us than RAM.